File: utility/file-viewer-main/apps/viewer-demo/public/example/code.js
const RENDERER_GROUPS = {
  document: ['doc', 'docx', 'pdf', 'ofd', 'typ', 'typst'],
  sheet: ['xlsx', 'xls', 'csv', 'ods'],
  drawing: ['dxf', 'dwg', 'dwf', 'dwfx', 'xps'],
  source: ['js', 'ts', 'tsx', 'vue', 'json', 'yaml']
}

const DEFAULT_ICON = {
  label: 'FILE',
  tone: '#64748b'
}

function normalizeExtension(filename = '') {
  const cleanName = filename.split('?')[0].split('#')[0]
  const dotIndex = cleanName.lastIndexOf('.')
  return dotIndex >= 0 ? cleanName.slice(dotIndex + 1).toLowerCase() : ''
}

export function resolvePreviewMeta(filename) {
  const extension = normalizeExtension(filename)

  for (const [group, extensions] of Object.entries(RENDERER_GROUPS)) {
    if (extensions.includes(extension)) {
      return {
        group,
        extension,
        icon: extension.toUpperCase(),
        supported: true
      }
    }
  }

  return {
    group: 'unknown',
    extension,
    icon: DEFAULT_ICON.label,
    supported: false
  }
}

export async function createPreviewFile(url, fallbackName = 'download.bin') {
  const response = await fetch(url, { credentials: 'include' })
  if (!response.ok) {
    throw new Error(`Download failed with ${response.status}`)
  }

  const blob = await response.blob()
  const name = decodeURIComponent(url.split('/').pop() || fallbackName)
  return new File([blob], name, { type: blob.type })
}

const samples = ['invoice.ofd', 'report.typ', 'drawing.dxf', 'component.vue', 'archive.unknown']
const summary = samples.map(resolvePreviewMeta)

console.table(summary)

File: utility/file-viewer-main/apps/viewer-demo/public/example/code.py
from dataclasses import dataclass
from pathlib import Path
from typing import Iterable


@dataclass(frozen=True)
class PreviewFile:
    name: str
    size: int
    source: str = "url"

    @property
    def extension(self) -> str:
        return Path(self.name).suffix.removeprefix(".").lower()

    @property
    def is_large(self) -> bool:
        return self.size > 5 * 1024 * 1024


RENDERERS = {
    "pdf": "pdfjs-dist",
    "ofd": "DLTech21/ofd.js",
    "typ": "typst.ts",
    "typst": "typst.ts",
    "dxf": "@flyfish-dev/cad-viewer",
    "dwg": "@flyfish-dev/cad-viewer",
    "dwf": "@flyfish-dev/cad-viewer",
    "dwfx": "@flyfish-dev/cad-viewer",
    "xps": "@flyfish-dev/cad-viewer",
    "json": "highlight.js",
    "py": "highlight.js",
}


def build_preview_queue(files: Iterable[PreviewFile]) -> list[dict[str, str]]:
    queue = []
    for item in files:
        renderer = RENDERERS.get(item.extension, "fallback")
        queue.append({
            "name": item.name,
            "renderer": renderer,
            "priority": "low" if item.is_large else "normal",
        })
    return queue


samples = [
    PreviewFile("invoice.ofd", 428000),
    PreviewFile("report.typ", 18000),
    PreviewFile("house.dwfx", 17_000_000),
    PreviewFile("trace.log", 12000),
]

for job in build_preview_queue(samples):
    print(f"{job['name']}: {job['renderer']} ({job['priority']})")
